java/arcs/core/analysis/ExpressionClaimDeducer.kt

```diff
@@ -0,0 +1,142 @@
+    expr.left.accept(this, ctx) + expr.right.accept(this, ctx)
+
+  override fun <T> visit(expr: Expression.FieldExpression<T>, ctx: Unit) =
+    when (val lhs = expr.qualifier) {
+      is Expression.FieldExpression<*> -> {
+        val base = lhs.accept(this, ctx)
+        Deduction(
+          base.scope,
+          base.context.mergeTop(Deduction.Analysis.Path(expr.field)),
+          base.aliases
+        )
+      }
+      is Expression.NewExpression -> {
+        val base = lhs.accept(this, ctx)
+        Deduction(
+          base.scope.lookup(expr.field),
+          base.context + Deduction.Analysis.Paths(listOf(expr.field)),
+          base.aliases
+        )
+      }
+      null -> Deduction(context = Deduction.Analysis.Paths(listOf(expr.field)))
+      else -> lhs.accept(this, ctx) + Deduction(
+        context = Deduction.Analysis.Paths(listOf(expr.field))
+      )
+    }
+
+  override fun <T> visit(expr: Expression.QueryParameterExpression<T>, ctx: Unit): Deduction {
+    TODO("Not yet implemented")
+  }
+
+  override fun visit(expr: Expression.NumberLiteralExpression, ctx: Unit) = Deduction()
+
+  override fun visit(expr: Expression.TextLiteralExpression, ctx: Unit) = Deduction()
+
+  override fun visit(expr: Expression.BooleanLiteralExpression, ctx: Unit) = Deduction()
+
+  override fun visit(expr: Expression.NullLiteralExpression, ctx: Unit) = Deduction()
+
+  override fun visit(expr: Expression.FromExpression, ctx: Unit) =
+    (expr.qualifier?.accept(this, ctx) ?: Deduction()) + Deduction(
+      aliases = mapOf(expr.iterationVar to expr.source.accept(this, ctx).context)
+    )
+
+  override fun visit(expr: Expression.WhereExpression, ctx: Unit): Deduction {
+    TODO("Not yet implemented")
+  }
+
+  override fun <T> visit(expr: Expression.SelectExpression<T>, ctx: Unit): Deduction {
+    val qualifier = expr.qualifier.accept(this, ctx)
+    val selection = expr.expr.accept(this, ctx)
+
+    val pathsWithAliases = selection.context.substitute(qualifier.aliases)
+
+    return qualifier + Deduction(
+      scope = selection.scope.substitute(qualifier.aliases),
+      context = when (expr.expr) {
+        is Expression.FieldExpression<*> -> {
+          Deduction.Analysis.Paths(Deduction.Analysis.Equal(pathsWithAliases.getPath()))
+        }
+        is Expression.BinaryExpression<*, *, *> -> Deduction.Analysis.Paths(
+          pathsWithAliases.paths.map { Deduction.Analysis.Derive(it) }
+        )
+        else -> pathsWithAliases
+      },
+      aliases = selection.aliases
+    )
+  }
+
+  override fun visit(expr: Expression.LetExpression, ctx: Unit): Deduction {
+    TODO("Not yet implemented")
+  }
+
+  override fun <T> visit(expr: Expression.FunctionExpression<T>, ctx: Unit): Deduction {
+    TODO("Not yet implemented")
+  }
+
+  /** Associates subexpressions and fields as a Derivation [Claim]. */
+  override fun visit(expr: Expression.NewExpression, ctx: Unit): Deduction =
```

Comment:
Question for the reviewer: What is a good way to simplify this logic? I'm open to suggestions, and will think of approaches in the meantime.